katdal/categorical.py

```diff
@@ -580,8 +580,70 @@ def concatenate_categorical(split_data, **kwargs):
     return data
 
 
-def sensor_to_categorical(sensor_timestamps, sensor_values, dump_midtimes, dump_period,
-                          greedy_values=None, initial_value=None, transform=None, allow_repeats=False, **kwargs):
+def _single_event_per_dump(events, greedy):
+    """Ensure that each dump is associated with a single sensor event.
+
+    This generates a sequence of cleaned-up sensor events (represented by
+    indices into the original `events` sequence), which ensures that each dump
+    is associated with a single event. When there are multiple events inside
+    a dump, pick the final one. In addition, some sensor values designated as
+    "greedy" will override non-greedy ones and grab a dump even if it is not
+    the final value. In this scenario, move the final (non-greedy) event to the
+    next dump by modifying its dump index in the `events` parameter. The
+    generator returns up to *N* events but not the special terminal event.
+
+    Parameters
+    ----------
+    events : mutable sequence of non-negative ints, length *N* + 1
+        Monotonic sequence of dump indices associated with each sensor event.
+        The last event is one past the last dump (i.e. the total number of
+        dumps). Be aware that this parameter is mutated by the function.
+    greedy : sequence of bool, length *N*
+        Flags indicating whether the sensor value at a given event is "greedy"
+
+    Yields
+    ------
+    event_index : non-negative int
+        Index into `events` sequence of the next cleaned up event (does not
+        yield the one-past-last-dump terminal event)
+
+    """
+    # The previous winning event is the dominant event in the previous dump
+    previous_winning_event = 0
+    previous_dump = 0
+    # This generates consecutive event indices with associated dump indices
+    for current_event, current_dump in enumerate(events):
+        # At the start of a new dump, process the events of the previous dump
+        if current_dump > previous_dump:
+            # This previous event segment is assumed to straddle dump boundary
+            assert current_event >= 1, "First sensor event not at dump 0"
+            event_at_dump_start = current_event - 1
+            # The normal victory condition is to be the final event in the dump
+            if not greedy[previous_winning_event]:
+                previous_winning_event = event_at_dump_start
+            winning_dump = events[previous_winning_event]
+            # Only yield winning event in immediate past to avoid duplicates
+            if previous_dump <= winning_dump < current_dump:
```

Comment:
Good question. The other `yield` further down has a need for the dump to be in the past, maybe I included it here too. The `event_at_dump_start` should be associated with `previous_dump` and `previous_winning_event` should not be ahead of it. So I think you are right that it is belt-and-braces.